contracts/azorius/Azorius.sol

```diff
@@ -0,0 +1,463 @@
+    bytes32 public constant DOMAIN_SEPARATOR_TYPEHASH =
+        0x47e79534a245952e8b16893a336b85a3d9ea9fa8c573f3d803afb92a79469218;
+    bytes32 public constant TRANSACTION_TYPEHASH =
+        0x72e9670a7ee00f5fbf1049b8c38e3f22fab7e9b85029e85cf9412f17fdd5c2ad;
+    uint256 public totalProposalCount; // Total number of submitted proposals
+    uint256 public timelockPeriod; // Delay between a proposal is passed and can be executed
+    address internal constant SENTINEL_STRATEGY = address(0x1);
+    mapping(uint256 => Proposal) internal proposals; // Proposals by proposal ID
+    mapping(address => address) internal strategies;
+
+    event ProposalCreated(
+        address strategy,
+        uint256 proposalId,
+        address proposer,
+        Transaction[] transactions,
+        string metadata
+    );
+    event TransactionExecuted(uint256 proposalId, bytes32 txHash);
+    event TransactionExecutedBatch(uint256 startIndex, uint256 endIndex);
+    event ProposalTimelocked(uint256 proposalId, uint256 timelockDeadline);
+    event AzoriusSetup(
+        address indexed initiator,
+        address indexed owner,
+        address indexed avatar,
+        address target
+    );
+    event EnabledStrategy(address strategy);
+    event DisabledStrategy(address strategy);
+    event TimelockPeriodUpdated(uint256 newTimelockPeriod);
+
+    constructor(
+        address _owner,
+        address _avatar,
+        address _target,
+        address[] memory _strategies,
+        uint256 _timelockPeriod
+    ) {
+        bytes memory initParams = abi.encode(
+            _owner,
+            _avatar,
+            _target,
+            _strategies,
+            _timelockPeriod
+        );
+        setUp(initParams);
+    }
+
+    function setUp(bytes memory initParams) public override initializer {
+        (
+            address _owner,
+            address _avatar,
+            address _target,
+            address[] memory _strategies,
+            uint256 _timelockPeriod
+        ) = abi.decode(initParams, (address, address, address, address[], uint256));
+        __Ownable_init();
+        avatar = _avatar;
+        target = _target;
+        setupStrategies(_strategies);
+        transferOwnership(_owner);
+        _updateTimelockPeriod(_timelockPeriod);
+
+        emit AzoriusSetup(msg.sender, _owner, _avatar, _target);
+    }
+
+    /// @notice Enables a voting strategy that can vote on proposals, only callable by the owner
+    /// @param _strategy Address of the strategy to be enabled
+    function enableStrategy(address _strategy) public onlyOwner {
+        require(
+            _strategy != address(0) && _strategy != SENTINEL_STRATEGY,
+            "Invalid strategy"
+        );
+        require(
+            strategies[_strategy] == address(0),
+            "Strategy already enabled"
+        );
+
+        strategies[_strategy] = strategies[SENTINEL_STRATEGY];
+        strategies[SENTINEL_STRATEGY] = _strategy;
```

Comment:
In `setupStrategies` --

```
        require(
            strategies[SENTINEL_STRATEGY] == address(0),
            "setupStrategies has already been called"
        );
```

It checks to make sure that `strategies[SENTINEL_STRATEGY] == address(0)` --- but after calling `enableStrategy`, wouldn't that update `strategies[SENTINEL_STRATEGY]` to not be address 0, thus allowing `setupStrategies` to be called again?

`setupStrategies` is only called by `setUp`, so maybe the check in `setupStrategies` is only guarding against `setUp` being called multiple times?

Reply:
Hmm so once `enableStrategy` is called, `strategies[SENTINEL_STRATEGY]` will no longer be equal to `address(0)`, so then `setupStrategies` will not be able to be called again, right?

Reply:
Oh yeah I read this wrong. Have been writing custom errors where the conditional logic is the inverse of "require" which is the cause.  All is good here, can ignore the comment.